### PR TITLE
r9 improvements

### PR DESCRIPTION
**Pull request by jlpoltrack**
To address: https://github.com/olliw42/mLRS-docu/issues/114

General notes:

- ELRS bootloader LED pattern added
- Recommendation to use FC to flash if there are USB<>UART issues.  I wonder if this should be the recommended way to flash as surely everyone using mLRS will be using a FC.
- Dip switch photos added
- Note for python3 prerequisite
- Reworked section on the serial port to recommend ESP32 + bridge sketch

### CODE REVIEW

**Comment by olliw42** [comment]
very nice !!!

@brad112358 your input on this would be very much appreciated

the pictures are great but I think it's not as illustrative as it could be, with the embedding in bullet points as now. Maybe they both could be put ion one pic, to the left and right, with some annotation on the left and right part. Would slim it down and make the bullets supefluous. Just a thought.

File: docs/FRSKY_R9.md [hunk 1]
```diff
@@ -27,11 +27,16 @@ Note that R9 hardware cannot connect with other mLRS boards which support 868/91
...
 
-1. The serial port of the R9M module can be configured to work either as "serial" or "CLI". This is done by setting the dip switch 1 (left dip switch): off (switch down) = CLI, on (switch up) = serial. Note that the dip switch position is read only at power up, i.e., one needs to re-power the module to make any change effective.
+1. The serial port of the R9M module can be configured to work either as "serial" or "CLI". This is done by setting the dip switch 1 (left dip switch): off (switch down) = CLI, on (switch up) = serial. Note that the dip switch position is read only at power up, i.e., one needs to re-power the module to make any change effective.  Refer to the photos below:
 
-2. One needs to somehow (un)invert the serial signals for most uses. You can buy or build a "Frsky inverter" dongle to connect a standard serial adapter.
+- CLI, dip switch 1 down:  
+<img src="images/Frsky_R9M_Dip1_Down.png" width="360px">
+- Serial, dip switch 1 up:
```
Comment (olliw42) [inline docs/FRSKY_R9.md:34]: the first is displayed as bullet and this one as dash ...

**Comment by jlpoltrack** [comment]
> the pictures are great but I think it's not as illustrative as it could be, with the embedding in bullet points as now. Maybe they both could be put ion one pic, to the left and right, with some annotation on the left and right part. Would slim it down and make the bullets supefluous. Just a thought.

Good point, added a new photo with annotations which replaces the clunky text.

**Comment by olliw42** [comment]
much nicer :)
I've fooled around with your latest pic (faking it)

dips

**Comment by jlpoltrack** [comment]
Dip switch annotations now in green for extra contrast :)

**Comment by olliw42** [comment]
👍 👍 👍 

let's hear @brad112358 for the elrs etc details

File: docs/FRSKY_R9.md [hunk 2]
```diff
@@ -240,17 +242,19 @@ Wiring the receiver to the JR Bay on the back of your radio and flashing the ELR
...
+6. When the python script reports "attempting to reboot into bootloader", power up the receiver by connecting the VCC wire to the 5 volt output of your serial adapter or a 5 volt power supply.  You should see the script report sync and begin the firmware download. If it fails, try again, the timing can be a bit tight. The first time you flash the .elrs file, you may not need to delay connecting the VCC wire.  If you have trouble getting the timing correct, you can hold down the button when powering up the receiver and the ELRS bootloader will keep running and not start the application code. The red and green LEDs will blink alternately while the receiver is in the ELRS bootloader.
```
Comment (olliw42) [inline docs/FRSKY_R9.md:257]: With regards to the last sentence, I think I could be made stronger as regards the potential use of this information. Like e.g. "The red and green LEDs will blink alternately while the receiver is in the ELRS bootloader, which allows visual check if the step was succesful and the bootloader is running.". Or anything along this line.

File: docs/FRSKY_R9.md [hunk 3]
```diff
@@ -27,11 +27,13 @@ Note that R9 hardware cannot connect with other mLRS boards which support 868/91
...
 
-2. One needs to somehow (un)invert the serial signals for most uses. You can buy or build a "Frsky inverter" dongle to connect a standard serial adapter.
+<img src="images/Frsky_R9M_Dip.png" width="720px">
 
-3. If you use the mLRS Lua configuration script for configuration and thus don't need the CLI for configuration, you can avoid the inverter dongle when you use one of the supported ESP32 boards with the mlrs-wireless-bridge sketch to connect via MAVLink wirelessly to a Ground Control Station. The ESP32 can directly use the inverted serial signals.
+2. Dealing with the inverted TTL signals is best addressed by using a seperate ESP32 module connected to the serial port as the ESP32 supports inverted TTL signals. The mlrs-wireless-bridge sketch will allow one to connect a Ground Control Station wirelessly via MAVLink. Additionally, the CLI can be accessed when using the mlrs-wireless-bridge sketch in Bluetooth mode.
```
Comment (olliw42) [inline docs/FRSKY_R9.md:34]: maybe along "... the CLi can be accessed via any ot the common terminal programs when ..."

File: docs/FRSKY_R9.md (hunk 2 above)
Comment (jlpoltrack) [reply docs/FRSKY_R9.md]: Reworked the last two sentences to make more sense.

File: docs/FRSKY_R9.md (hunk 3 above)
Comment (jlpoltrack) [reply docs/FRSKY_R9.md]: Sure, added.

File: docs/FRSKY_R9.md [hunk 4]
```diff
@@ -240,17 +242,19 @@ Wiring the receiver to the JR Bay on the back of your radio and flashing the ELR
 
 After flashing the ELRS bootloader, you can connect the serial port as described in the link above and use the ELRS UARTupload.py script to flash/update mLRS with latest mLRS firmware .elrs file. There are, of course, separate .elrs image files for each receiver:
 
+*Note: Python3 is a prerequisite to use the ELRS UARTUpload.py script.*
```
Comment (olliw42) [inline docs/FRSKY_R9.md:245]: why this in italics unlike all other comments? All italics?

File: docs/FRSKY_R9.md [hunk 5]
```diff
@@ -240,17 +242,19 @@ Wiring the receiver to the JR Bay on the back of your radio and flashing the ELR
...
+
 1. Download the latest mLRS firmware or build with MLRS\_FEATURE\_ELRS\_BOOTLOADER defined.
 
 2. Download or git clone the [ExpressLRS](https://github.com/ExpressLRS/ExpressLRS) repository. Since you only need the src/python folder, if you have subversion installed, you could use the svn github interface to get just what you need. From the command line type something like this: "svn export https://github.com/ExpressLRS/ExpressLRS/trunk/src/python"
 
-3. Connect a TTL USB serial adapter to your computer and install a driver if it's not recognized. If you don't have a TTL USB serial adapter, you should be able to use your flight controller. See step 8.
+3. Connect a TTL USB serial adapter to your computer and install a driver if it's not recognized.  If you're having trouble flashing or don't have a TTL USB serial adapter, you should be able to use your flight controller - see step 8.
```
Comment (olliw42) [inline docs/FRSKY_R9.md:251]: does this really belong to here? Should it be a Note in this bullet, for better visibility?

File: docs/FRSKY_R9.md (hunk 4 above)
Comment (jlpoltrack) [reply docs/FRSKY_R9.md]: Italics removed.

File: docs/FRSKY_R9.md (hunk 5 above)
Comment (jlpoltrack) [reply docs/FRSKY_R9.md]: This is now a note but I wonder if flashing via FC should be the recommended first way to try as opposed to recommending to flash via USB<>UART first.

Comment (brad112358) [reply docs/FRSKY_R9.md]: I think for new users, flashing via USB serial is easier.  At least until I get around to enhancing the python script to eliminate the manual passthrough setup in ardupilot.

**Comment by olliw42** [comment]
from my side lgtm

not sure what @brad112358 has to say, especially on teh ELRS stuff

**Comment by brad112358** [comment]
Thanks for doing this!  Updating this has been on my list of things to do for a while.  I think the only item on my list that you didn't address here is the choice of esp32 modules.  The document sort of suggests that the M5Stamp C3U Mate is the best choice, but when I wrote this, I didn't realize that the esp32 chips that support BT LE don't still support classic BT serial.  The M5 Stamp Pico Mate can do both Bluetooth serial and WiFi so it may be a better choice for many new users in spite of needing a special serial adapter to program it.  I've been meaning to add this info and add the photos of the M5Stamp Pico that I included in the Ardupilot documentation page.

**Comment by olliw42** [comment]
that's a very good point indeed
it needs to make better clear that not all ESP32 modules suitable for mlrs-wirelessbridge are also BT classic capable (it nowadays is probably fair to call these just a small subset), but that this is required for "easier" CLI access...

**Comment by brad112358** [comment]
I think the newer ESP32 modules can still be used for CLI access, via telnet or a TCP terminal program when TCP is selected for the sketch.  I tested the a wifi terminal app on android, but telnet should also work fine on windows or linux or android.  Just specify the matching port when starting telnet.  (I still need to test telnet).

**Comment by jlpoltrack** [comment]
Expanded the section on how to use ESP32 + CLI.

**Comment by brad112358** [comment]
It all looks good to me!  I'll try to add a photo showing the connector on the M5 Stamp Pico Mate along with a link on flashing it at some point.  It may also someday be possible to do serial via BLE 5.0, but I think work is required on the ground station side.

**Comment by jlpoltrack** [comment]
@brad112358 Alright to handle image update in separate PR?

@olliw42 If brad agrees to the above then I will merge.

**Comment by brad112358** [comment]
>  @brad112358 Alright to handle image update in separate PR?

Yes.  That's what I was trying to say.    :)

**Pull request merged**